### DISCUSSION TITLE
How to order by count of another realtion efficiently without using query builder?

### DISCUSSION THREAD

**Discussion by squivix** [mikro-orm/mikro-orm#4540] [Q&A] [answered]
Say I have 2 entties: `EntityA` and `EntityB`. There's a M:N relationship between them. I need to fetch all of the `EntityA` records, ordered by how many `EntityB`s are is associated with them. How to do this with the EM ideally without using the query builder?

Right now I'm using a formula for coutning entity_bs:
```ts
    @Formula((alias: string) => `SELECT COUNT(DISTINCT pivot_entity_b_entity_a.entity_b_id) FROM pivot_entity_b_entity_a WHERE pivot_entity_b_entity_a.entity_a_id=${alias}.id)`, {
        type: "number"
    })
    entityBCount?: number;
```

And then I order by that formula field:
```ts
await em.find(EntityA, {}, {orderBy: {entityBCount:"asc"}})
```

I went with formulas after the discussion in #4110 

But the generated query uses subqueries in the orderBy clause, which is very slow. It looks roughly like this:
```sql
SELECT entity_a.*, 
SELECT COUNT(DISTINCT pivot_entity_b_entity_a.entity_b_id) FROM pivot_entity_b_entity_a WHERE pivot_entity_b_entity_a.entity_a_id=entity_a.id as entity_b__count
FROM entity_a
ORDER BY (SELECT COUNT(DISTINCT pivot_entity_b_entity_a.entity_b_id) FROM pivot_entity_b_entity_a WHERE pivot_entity_b_entity_a.entity_a_id=entity_a.id)) DESC;
```

I need it to be faster. Ideally using joins like this:

```sql
SELECT entity_a.*, COUNT(DISTINCT pivot_entity_b_entity_a.entity_b_id) AS entity_bs_count
FROM entity_a
         LEFT JOIN pivot_entity_b_entity_a ON entity_a.id = pivot_entity_b_entity_a.entity_a_id
GROUP BY entity_a.id
ORDER BY entity_bs_count DESC;
```

But at least don't repeat the formula in the orderBy when it's already been selected. Just use its alias:
```sql
SELECT entity_a.*, 
SELECT COUNT(DISTINCT pivot_entity_b_entity_a.entity_b_id) FROM pivot_entity_b_entity_a WHERE pivot_entity_b_entity_a.entity_a_id=entity_a.id as entity_b_count
FROM entity_a
ORDER BY (entity_b_count) DESC;
```

**Comment by B4nan** [comment, marked as answer]
You will need a query builder if you want to have control over how the query looks (or a raw sql, no shame in doing that, especially if you already have it).

Maybe you could define the pivot entity, with a formula for the `COUNT(DISTINCT pivot_entity_b_entity_a.entity_b_id)` and try to do it through that? But it feels a bit overcomplicated.

**Reply by squivix** [reply]
Ok I'll go with raw queries.

How do I alias my columns so that `em.map()` recognizes them?

Right now I'm getting an error: `Collection<EntityB> of entity EntityA[0] is not initialized `

**Reply by B4nan** [reply]
There are no extra steps for the root entity. You can't really map anything else, for relations that are fetch joined you'd have to use the QB as its not just about having correctly aliased fields, you also have to provide the information about what is joined under what alias (and there are no public APIs for that anyway).

What is the stack trace? Provide some code.

**Reply by squivix** [reply]
Oh so `em.map()` doesn't initialize collections for relations. Ok guess I'll learn to use the query builder. If only there was an easy way to convert SQL to qb :-|


No need for a stack trace. The error is thrown from my own code when I call `entityA.entityBs.getItems()` and the collection is not initialized. I thought it would be.

**Reply by B4nan** [reply]
> Oh so em.map() doesn't initialize collections for relations.

It does, but not from flat results produced by `qb.execute()`, and not without additional metadata - which is both handled if you use `qb.getResult()`. If you would have it grouped to a structure like `{ name: '...', coll1: [{ foo: 1 }, { foo: 2 }] })`, it would hydrate the collection items too. 

But in your query, you only select the root entity fields plus that count (`SELECT entity_a.*, COUNT(DISTINCT pivot_entity_b_entity_a.entity_b_id) AS entity_bs_count`), so why you would expect that to populate some relations that are not part of the select clause?

If you use QB with the `joinAndSelect` method, it will handle aliasing (so there can't be any conflicts, e.g. when the same column name exists in multiple joined tables), and keep metadata about this for later, which is used to properly map the results to a deep entity graph. See how the `QB.getResultList` is implemented (which is alias for `qb.getResult` and is the thing that gets executed if you await the QB instance directly). It also uses `qb.execute`, but then uses `driver.mergeJoinedResult` to merge the results to the expected structure (this is where the aliased columns are resolved and grouped), before passing it to the `em.map`. It also handles propagation of populate hints, which are needed for proper implicit serialization.

https://github.com/mikro-orm/mikro-orm/blob/master/packages/knex/src/query/QueryBuilder.ts#L639

**Reply by squivix** [reply]
I just found out that `raw()` and `sql.ref()` helpers will be coming to v6 and can be used in orderBy (discussion #4503)

Will that make this possible without using the query builder?
